texar/tf/utils/dynamic_decode.py

```diff
@@ -0,0 +1,314 @@
+        """
+
+        (next_outputs, sample_ids, logits, state) = decoder.update(
+            inputs, state, time)
+        reach_max = tf.equal(time+1, decoder.max_decoding_length)
+        (decoder_finished, next_inputs, decoder_state) = tf.cond(
+            reach_max,
+            lambda: (tf.cast(tf.ones(tf.shape(sample_ids)[0]), tf.bool),
+                     decoder._helper.start_inputs,
+                     state),
+            lambda: decoder.next_inputs(sample_ids, time, logits, state)
+        )
+        if decoder.tracks_own_finished:
+            next_finished = decoder_finished
+        else:
+            next_finished = tf.logical_or(decoder_finished, finished)
+        next_sequence_lengths = tf.where(
+            tf.logical_not(finished),
+            tf.fill(tf.shape(sequence_lengths), time + 1),
+            sequence_lengths)
+
+        nest.assert_same_structure(state, decoder_state)
+        nest.assert_same_structure(outputs_ta, next_outputs)
+        nest.assert_same_structure(inputs, next_inputs)
+
+        # Zero out output values past finish
+        impute_finished=True
+        if impute_finished:
+            emit = nest.map_structure(
+                lambda out, zero: tf.where(finished, zero, out),
+                next_outputs,
+                zero_outputs)
+        else:
+            emit = next_outputs
+
+        # Copy through states past finish
+        def _maybe_copy_state(new, cur):
+            # TensorArrays and scalar states get passed through.
+            if isinstance(cur, tf.TensorArray):
+                pass_through = True
+            else:
+                new.set_shape(cur.shape)
+                pass_through = (new.shape.ndims == 0)
+            return new if pass_through else tf.where(finished, cur, new)
+
+        if impute_finished:
+            next_state = nest.map_structure(
+                _maybe_copy_state, decoder_state, state)
+        else:
+            next_state = decoder_state
+
+        outputs_ta = nest.map_structure(lambda ta, out: ta.write(time, out),
+                                        outputs_ta, emit)
+        return (time + 1, outputs_ta, next_state, next_inputs, next_finished,
+                next_sequence_lengths)
+
+    res = tf.while_loop(
+        condition,
+        body,
+        loop_vars=(
+            initial_time,
+            initial_outputs_ta,
+            initial_state,
+            initial_inputs,
+            initial_finished,
+            initial_sequence_lengths,
+        ),
+        maximum_iterations=maximum_iterations)
+
+    final_outputs_ta = res[1]
+    final_state = res[2]
+    final_sequence_lengths = res[5]
+
+    final_outputs = nest.map_structure(lambda ta: ta.stack(), final_outputs_ta)
+
+    try:
+        final_outputs, final_state = decoder.finalize(
+            final_outputs, final_state, final_sequence_lengths)
+    except NotImplementedError:
+        pass
+
+    def _transpose_batch_time(x):
```

Comment:
Do we have a `util` method for such functionality? I feel that this is a commonly used function. @TomNong